src/utils/point_locator_nanoflann.C

```diff
@@ -0,0 +1,427 @@
+{
+  LOG_SCOPE("init()", "PointLocatorNanoflann");
+
+  if (!_initialized)
+    {
+      // If _master == nullptr, then we _are_ the master, and thus
+      // responsible for initializing.
+      bool we_are_master = (_master == nullptr);
+
+      // If we are the master PointLocator, fill in the _point_cloud
+      // data structure with active, local element centroids.
+      if (we_are_master)
+        {
+          _elems = std::make_shared<std::vector<Elem *>>();
+          _point_cloud = std::make_shared<std::vector<Point>>();
+
+          // Make the KD-Tree out of active element centroids.
+          //
+          // Note: we use active elements rather than active+local
+          // elements since we sometimes need to be able to locate
+          // points in ghosted elements (for example in Periodic BCs).
+          // Active elements are also natural to use in the
+          // DistributedMesh case. The size of the KD-Tree on each
+          // processor therefore scales with the number of elements on
+          // each processor in either case.
+          //
+          // Note 2: the approximate amount of space we should reserve
+          // here is going to be different for ReplicatedMesh
+          // (n_active_elem()) vs. DistributedMesh (n_active_local_elem())
+          // so we just let the containers resize themselves
+          // automatically.
+          for (const auto & elem : _mesh.active_element_ptr_range())
+            {
+              _elems->push_back(elem);
+              _point_cloud->push_back(elem->centroid());
+            }
+
+          // Construct the KD-Tree
+          _kd_tree = std::make_shared<kd_tree_t>
+            (LIBMESH_DIM, *this, nanoflann::KDTreeSingleIndexAdaptorParams(/*max leaf=*/10));
+
+          _kd_tree->buildIndex();
+        }
+      else // we are not master
+        {
+          // Cast the _master to the appropriate type, and point to
+          // its data arrays.
+          const auto my_master =
+            cast_ptr<const PointLocatorNanoflann *>(this->_master);
+
+          // Point our data structures at the master's
+          _elems = my_master->_elems;
+          _point_cloud = my_master->_point_cloud;
+          _kd_tree = my_master->_kd_tree;
+        }
+
+      // We are initialized now
+      this->_initialized = true;
+    }
+}
+
+nanoflann::KNNResultSet<Real>
+PointLocatorNanoflann::kd_tree_find_neighbors(const Point & p,
+                                              std::size_t num_results) const
+{
+  // We are searching for the Point(s) closest to Point p.
+  //
+  // TODO: The kd_tree's findNeighbors() routine needs a pointer to
+  // Real of length LIBMESH_DIM. It might be convenient if libMesh
+  // Points had a data() member that provided this, for now we just
+  // copy the coordinates into a std::array of the appropriate size.
+  std::array<Real, LIBMESH_DIM> query_pt;
+  for (int i=0; i<LIBMESH_DIM; ++i)
+    query_pt[i] = p(i);
```

Comment:
We can't initialize that in one line?

Reply:
Yeah with some `#ifdefs` and carefully placed commas we could, but it seemed shorter and cleaner to just do this loop.

Reply:
I think the assembly code should be almost identical after unrolling and optimizations happen, but I did not check this.

Reply:
Oh, of course I forgot about the LIBMESH_DIM < 3 cases.  Yeah, this is fine; I was just hoping to save lines of code, not FLOPs.